src/middlewared/middlewared/plugins/nis.py

```diff
@@ -0,0 +1,189 @@
+        If for some reason, the cache entry indicating Directory Service state
+        does not exist, re-run a status check to generate a key, then return it.
+        """
+        nis = await self.middleware.call('nis.config')
+        if not nis['enable']:
+            return 'DISABLED'
+        else:
+            try:
+                return (await self.middleware.call('cache.get', 'NIS_State'))
+            except KeyError:
+                await self.started()
+                return (await self.middleware.call('cache.get', 'NIS_State'))
+
+    @private
+    async def start(self):
+        """
+        Refuse to start the service if a state has already been set in the cache. This is to prevent multiple
+        simultaneous service starts. Even in faulted state, it is safer to cleanly stop the service before starting.
+        Even in a faulted state, it is safer to cleanly stop the service before restarting.
+        """
+        state = await self.get_state()
+        if state == 'FAULTED':
+            raise CallError(f'NIS service is currently in faulted state. Review logs for error message or try clean restart of NIS service.', errno.EBUSY)
+
+        if state == 'HEALTHY':
+            raise CallError(f'NIS service is currently in healthy state. Valid command options are "restart" or "reload"', errno.EBUSY)
+
+        if state != 'DISABLED':
+            raise CallError(f'Current state of NIS service is: [{state}]. Wait until operation completes.', errno.EBUSY)
+
+        await self.middleware.call('nis.update', {'enable': True})
+        await self.__set_state(DSStatus['JOINING'])
+        await self.middleware.call('etc.generate', 'rc')
+        await self.middleware.call('etc.generate', 'pam')
+        await self.middleware.call('etc.generate', 'hostname')
+        await self.middleware.call('etc.generate', 'nss')
+        nis = await self.middleware.call('nis.config')
+        setnisdomain = await run(['/bin/domainname', nis['domain']], check=False)
+        if setnisdomain.returncode != 0:
+            await self.__set_state(DSStatus['FAULTED'])
+            raise CallError(f'Failed to set NIS Domain to [{nis["domain"]}]: {setnisdomain.stderr.decode()}')
+
+        ypbind = await run(['/usr/sbin/service', 'ypbind', 'onestart'], check=False)
+        if ypbind.returncode != 0:
+            await self.__set_state(DSStatus['FAULTED'])
+            raise CallError(f'ypbind failed: {ypbind.stderr.decode()}')
+
+        await self.__set_state(DSStatus['HEALTHY'])
+        await self.middleware.call('nis.cache_fill')
+        return True
+
+    @private
+    async def started(self):
+        """
+        If for some reason, the status check returns false, but NIS is enabled
+        set the NIS_state value to FAULTED.
+        The return code from ypwhich is not a reliable health indicator.
+        """
+        nis = await self.middleware.call('nis.config')
+        ypwhich = await run(['/usr/bin/ypwhich'], check=False)
```

Comment:
Didnt you mention ypwhich could block?

subprocess has a timeout param, if that doesnt work we can use asyncio.wait_for